### PR TITLE
DATAES-773 - Add search-as-you-type field support

### PR DESCRIPTION
**Pull request by alesharik**
This PR adds search-as-you-type field support

- [x] You have read the [Spring Data contribution guidelines](https://github.com/spring-projects/spring-data-build/blob/master/CONTRIBUTING.adoc).
- [x] There is a ticket in the bug tracker for the project in our [JIRA](https://jira.spring.io/browse/DATAES).
- [x] You use the code formatters provided [here](https://github.com/spring-projects/spring-data-build/tree/master/etc/ide) and have them applied to your changes. Don’t submit any formatting related changes.
- [x] You submit test cases (unit or integration tests) that back your changes.
- [x] You added yourself as author in the headers of the classes you touched. Amend the date range in the Apache license header if needed. For new types, add the license header (copy from another file and set the current year only).

### CODE REVIEW

**Comment by alesharik** [comment]
I re-implemented it, but now we have confusing `maxShingleSize` property with default value of 3, because we can't put nulls into annotation properties :/

File: src/main/java/org/springframework/data/elasticsearch/annotations/Field.java [hunk 1]
```diff
@@ -148,4 +149,9 @@
 	 * @since 4.0
 	 */
 	double scalingFactor() default 1;
+
+	/**
+	 * @since 4.0
+	 */
+	int maxShingleSize() default 3;
```
Comment (sothawo) [inline src/main/java/org/springframework/data/elasticsearch/annotations/Field.java]:
```suggestion
	int maxShingleSize() default -1;
```
we should use -1 here to denote a value that is not set (I forgot that we can't use null)

File: src/main/java/org/springframework/data/elasticsearch/annotations/InnerField.java [hunk 2]
```diff
@@ -117,4 +118,9 @@
 	 * @since 4.0
 	 */
 	double scalingFactor() default 1;
+
+	/**
+	 * @since 4.0
+	 */
+	int maxShingleSize() default 3;
```
Comment (sothawo) [inline src/main/java/org/springframework/data/elasticsearch/annotations/InnerField.java]:
```suggestion
	int maxShingleSize() default -1;
```

same as above

File: src/main/java/org/springframework/data/elasticsearch/core/index/MappingParameters.java [hunk 3]
```diff
@@ -88,6 +90,7 @@
 	private Similarity similarity = Similarity.Default;
 	private TermVector termVector = TermVector.none;
 	private double scalingFactor = 1.0;
+	private int maxShingleSize;
```
Comment (sothawo) [inline src/main/java/org/springframework/data/elasticsearch/core/index/MappingParameters.java]:
```suggestion
	private Integer maxShingleSize = null;
```

here we can have a nullable field

File: src/main/java/org/springframework/data/elasticsearch/core/index/MappingParameters.java [hunk 4]
```diff
@@ -136,6 +139,9 @@ private MappingParameters(Field field) {
 		similarity = field.similarity();
 		termVector = field.termVector();
 		scalingFactor = field.scalingFactor();
+		maxShingleSize = field.maxShingleSize();
```
Comment (sothawo) [inline src/main/java/org/springframework/data/elasticsearch/core/index/MappingParameters.java]:
```suggestion
		maxShingleSize = field.maxShingleSize() >= 0 ? field.maxShingleSize : null;
```
The check in the Assert should account for a not set default value of null

File: src/main/java/org/springframework/data/elasticsearch/core/index/MappingParameters.java [hunk 5]
```diff
@@ -165,6 +171,9 @@ private MappingParameters(InnerField field) {
...
+		Assert.isTrue(type != FieldType.Search_As_You_Type || (maxShingleSize >= 2 && maxShingleSize <= 4),
+				"maxShingleSize must be in inclusive range from 2 to 4 for field type search_as_you_type");
 	}
```
Comment (sothawo) [inline src/main/java/org/springframework/data/elasticsearch/core/index/MappingParameters.java]: see my comment in the other constructor

File: src/main/java/org/springframework/data/elasticsearch/core/index/MappingParameters.java [hunk 6]
```diff
@@ -269,5 +278,9 @@ public void writeTypeAndParametersTo(XContentBuilder builder) throws IOException
 		if (type == FieldType.Scaled_Float) {
 			builder.field(FIELD_PARAM_SCALING_FACTOR, scalingFactor);
 		}
+
+		if (maxShingleSize != 3) {
```
Comment (sothawo) [inline src/main/java/org/springframework/data/elasticsearch/core/index/MappingParameters.java]:
```suggestion
		if (maxShingleSize != null) {
```

**Comment by sothawo** [comment]
Your PR has been merged, thank you again for this. And sorry, that I didn't find the time to merge before the release of 4.0.0.RC1 this morning.

**Pull request merged**